### PR TITLE
Simplify Notify configuration mechanism

### PR DESCRIPTION
**Pull request by kevindew**
Trello: https://trello.com/c/G35ncFfw/645-simplify-configuration-for-email-sending

This replaces a number of concepts that are used to configure the Notify usage of Email Alert API. This has been done to reflect how the system is used in practice, to simplify the system and to resolve a problem where most emails weren't sent in integration and staging. The variety of configuration levers have been changed to now use a single `GOVUK_NOTIFY_RECIPIENTS` environment variable for deciding who'll get emails by Notify, those who are permitted will be sent an email with Notify and those that won't will have their email logged. The best way to review this is commit by commit.

Concepts changed:

- `provider` - this is now replaced by a conditional that decides whether email is sent to Notify or logged based on the value of `GOVUK_NOTIFY_RECIPIENTS`;
- `email_subject_prefix` - this feature is now removed, it appeared unnecessary as the senders are prefixed and it should be fully redundant when [switch integration/staging to send emails to a Google group](https://trello.com/c/l4ahPmzS/644-switch-integration-staging-to-send-emails-to-a-google-group) is completed;
- `email_address_override` - this feature wasn't used;
- `email_address_override_whitelist` - replaced by `GOVUK_NOTIFY_RECIPIENTS` environment variable;
- `email_address_override_whitelist_only` - no longer needed as an environment variable because an entry in `GOVUK_NOTIFY_RECIPIENTS` acts as an allowed list;
- `expect_status_update_callbacks` - this had no effect on the application anymore.

More information in the commits.

### CODE REVIEW

**Review by benthorner — APPROVED** [review]
Much better, and thanks for fixing a few other small things along the way 🎨. (Note to self: the new environment variable is [already in place](https://github.com/alphagov/govuk-puppet/pull/10859).)

Just a few minor comments. Also, it does seem a bit strange that the tests all send to `*` but the default is `""`. Perhaps the default should actually be `*`?

File: spec/integration/sending_email_to_a_recipient_spec.rb [hunk 1]
```diff
@@ -12,12 +12,12 @@
   let(:email_address_override_whitelist) { nil }
 
   def expect_notify_request_for(email_address:)
-    request = a_request(:post, /fake-notify/)
+    request = a_request(:post, /notifications\.service\.gov\.uk/)
```
[the remark this answers is not in the document]
Comment (kevindew) [reply spec/integration/sending_email_to_a_recipient_spec.rb]: You still need the escape with curly braces as it's to indicate the period doesn't mean any character.

File: app/services/send_email_service/send_pseudo_email.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  def initialize(email)
+    @email = email
+  end
+
+  def self.call(*args)
+    new(*args).call
+  end
+
+  def call
+    Rails.logger.info(<<~INFO)
```
[the remark this answers is not in the document]
Comment (kevindew) [reply app/services/send_email_service/send_pseudo_email.rb:11]: Yeah, I left the approach that was there before which I wasn't hugely familiar with. It's probably most idiomatic if I drop the braces from this call.

File: config/notify.yml [hunk 3]
```diff
@@ -1,18 +1,14 @@
 <% fake_api_key = "fake_api_key-00000000-0000-0000-0000-000000000000-00000000-0000-0000-0000-000000000000" %>
 <% fake_base_url = "http://fake-notify.com" %>
-<% template_id = "cb633abc-6ae6-4843-ae6f-82ca500b6de2" %>
```
[the remark this answers is not in the document]
Comment (kevindew) [reply config/notify.yml:3]: We used to use it in tests but no longer do. I'll amend the commit.

**Comment by kevindew** [comment]
> Much better, and thanks for fixing a few other small things along the way 🎨. (Note to self: the new environment variable is [already in place](https://github.com/alphagov/govuk-puppet/pull/10859).)
> 
> Just a few minor comments. Also, it does seem a bit strange that the tests all send to `*` but the default is `""`. Perhaps the default should actually be `*`?

Do you mean for test or all environments? It makes sense to be nil for all environments by default so that sending to Notify is something that has to be turned on rather than turned off. 

The default of `""` on https://github.com/alphagov/email-alert-api/pull/1489/files#diff-9e30a404f2ecf5367673492e503ceaa069c7ba93d34e9743bd0593d1e55d9df8R26 is just used since it makes it simpler to do the array conversion compared to a nil.

File: spec/integration/sending_email_to_a_recipient_spec.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply spec/integration/sending_email_to_a_recipient_spec.rb]: Oh yes, sorry 👍.

**Comment by benthorner** [comment]
> Do you mean for test or all environments? It makes sense to be nil for all environments by default so that sending to Notify is something that has to be turned on rather than turned off.

Yeah, I'm challenging whether it really matters that much, since the Integration/Staging accounts are restricted anyway. But it's only a minor thing I noticed when comparing the tests vs. the code - not fussed if it stays as-is.

**Comment by kevindew** [comment]
> > Do you mean for test or all environments? It makes sense to be nil for all environments by default so that sending to Notify is something that has to be turned on rather than turned off.
> 
> Yeah, I'm challenging whether it really matters that much, since the Integration/Staging accounts are restricted anyway. But it's only a minor thing I noticed when comparing the tests vs. the code - not fussed if it stays as-is.

Ok cool, maybe I'm missing a nuance but I think it makes more sense as-is so I'll keep it that way and get this merged. Thanks for being flexible 👍 .

**Pull request merged**